network/simple-onchain-discovery/src/builder.rs

```diff
@@ -71,6 +77,7 @@ impl ConfigurationChangeListenerBuilder {
         let config = self.config.take().expect("Listener must be configured");
         self.listener = Some(ConfigurationChangeListener::new(
             config.network_context,
+            config.expected_pubkey,
```

Comment:
can't comment on the exact line, but here:

```rust
    pub fn start(&mut self, executor: &Handle) -> &mut Self {
        assert_eq!(self.state, State::BUILT);
        self.state = State::STARTED;
        let listener = self.listener.take().expect("Listener must be built");
>>>>    executor.spawn(listener.start());    <<<<
        self
    }
```

you can see we're spawning a task `listener.start()` but not holding onto the `JoinHandle`. This means if the change listener panics, the task will print the error and shutdown, but the panic won't propagate and bring down the node. In fact, it will be worse, because the change listener will never restart and we'll just never get discovery updates ever again...

FWIW, you have the exact same problem if you use `std::thread::spawn` and just drop the `JoinHandle`; it won't propagate any panics crossing the thread boundary since you've detached the thread.

to be honest, our graceful shutdown process in diem-node... doesn't really exist : p

it seems the "proper" approach would be to look at each of the `executor::spawn` and `thread::spawn` points across the codebase, and decide "is a panic in this module recoverable?" if yes, it's probably fine to just drop the `JoinHandle` as we do already. in some places, we would probably want to restart the task. Otherwise, we need to hold onto the `JoinHandle`s and `std::panic::resume_unwind` the panics.

Reply:
Yeah, in this case we want to propagate the panic.

As a whole, I believe we're avoiding unwinding panics, and just taking a panic as a reason to restart the whole process.  Otherwise, the `panic!` -> `unwind` pattern is basically like using a try catch in another language, which I'm pretty sure rust was trying to avoid.

Reply:
Right, I would personally err on just propagating the panics everywhere, with the possible exception of a few isolated places like the `Peer` actor or transport upgrades, which is our first line-of-defense against untrusted input. Most of all, catching panics makes it a lot harder for our test infra to work properly, since it hides most latent bugs.